staging/vhost-device-can/src/vhu_can.rs

```diff
@@ -0,0 +1,765 @@
+            let descriptors: Vec<_> = desc_chain.clone().collect();
+
+            if descriptors.is_empty() {
+                warn!("Error::UnexpectedDescriptorCount");
+                return Err(Error::UnexpectedDescriptorCount(descriptors.len()));
+            }
+
+            trace!("descriptors.len(): {:?}", descriptors.len());
+
+            let desc_request = descriptors[0];
+            if desc_request.is_write_only() {
+                warn!("Error::UnexpectedWriteOnlyDescriptor");
+                return Err(Error::UnexpectedWriteOnlyDescriptor(0));
+            }
+
+            if desc_request.len() as usize != size_of::<VirtioCanCtrlRequest>() {
+                trace!("UnexpectedDescriptorSize, len = {:?}", desc_request.len());
+                return Err(Error::UnexpectedDescriptorSize(
+                    size_of::<VirtioCanCtrlRequest>(),
+                    desc_request.len(),
+                ));
+            }
+
+            let request = desc_chain
+                .memory()
+                .read_obj::<VirtioCanCtrlRequest>(desc_request.addr())
+                .map_err(|_| Error::DescriptorReadFailed)?;
+
+            match request.msg_type.into() {
+                VIRTIO_CAN_SET_CTRL_MODE_START => {
+                    //TODO: vcan->busoff = false;
+                    trace!("VIRTIO_CAN_SET_CTRL_MODE_START");
+                    Ok(())
+                }
+                VIRTIO_CAN_SET_CTRL_MODE_STOP => {
+                    //TODO: vcan->busoff = false;
+                    trace!("VIRTIO_CAN_SET_CTRL_MODE_STOP");
+                    Ok(())
+                }
+                _ => {
+                    trace!("Ctrl queue: msg type 0x{:?} unknown", request.msg_type);
+                    return Err(Error::HandleEventUnknown);
+                }
+            }?;
+
+            let desc_response = descriptors[1];
+            if !desc_response.is_write_only() {
+                return Err(Error::UnexpectedReadableDescriptor(1));
+            }
+
+            let response = VIRTIO_CAN_STATUS_OK;
+
+            desc_chain
+                .memory()
+                .write_slice(response.as_slice(), desc_response.addr())
+                .map_err(|_| Error::DescriptorWriteFailed)?;
+
+            if vring
+                .add_used(desc_chain.head_index(), desc_response.len())
+                .is_err()
+            {
+                warn!("Couldn't return used descriptors to the ring");
+            }
+        }
+
+        Ok(true)
+    }
+
+    fn process_tx_requests(
+        &self,
+        requests: Vec<CanDescriptorChain>,
+        vring: &VringRwLock,
+    ) -> Result<bool> {
+        trace!("process_tx_requests");
+
+        if requests.is_empty() {
+            trace!("requests.is_empty()");
+            return Ok(true);
+        }
+
+        for desc_chain in requests {
+            let descriptors: Vec<_> = desc_chain.clone().collect();
+
+            if descriptors.len() != 2 {
```

Comment:
I am not sure if we have to do it now but at some point we can use Reader/Writer modules instead of descriptors[] so the device is independent of how the guest build the request. It also reduces code.